lib/mayaUsd/ufe/UsdUndoConnectionCommands.cpp

```diff
@@ -284,6 +284,139 @@ UsdUndoDeleteConnectionCommand::Ptr UsdUndoDeleteConnectionCommand::create(
     return std::make_shared<UsdUndoDeleteConnectionCommand>(srcAttr, dstAttr);
 }
 
+bool canRemoveSrcProperty(const PXR_NS::UsdAttribute& srcAttr)
+{
+
+    // Do not remove if it has a value.
+    if (srcAttr.HasValue()) {
+        return false;
+    }
+
+    PXR_NS::SdfPathVector connectedAttrs;
+    srcAttr.GetConnections(&connectedAttrs);
+
+    // Do not remove if it has connections.
+    if (!connectedAttrs.empty()) {
+        return false;
+    }
+
+    const auto prim = srcAttr.GetPrim();
+
+    if (!prim) {
+        return false;
+    }
+
+    const auto primParent = prim.GetParent();
+
+    if (!primParent) {
+        return false;
+    }
+
+    // Do not remove if there is a connection with a prim.
+    for (const auto& childPrim : primParent.GetChildren()) {
+        if (childPrim != prim) {
+            for (const auto& attribute : childPrim.GetAttributes()) {
+                const PXR_NS::UsdAttribute dstUsdAttr = attribute.As<PXR_NS::UsdAttribute>();
+                if (isConnected(srcAttr, dstUsdAttr)) {
+                    return false;
+                }
+            }
+        }
+    }
+
+    // Do not remove if there is a connection with the parent prim.
+    for (const auto& attribute : primParent.GetAttributes()) {
+        const PXR_NS::UsdAttribute dstUsdAttr = attribute.As<PXR_NS::UsdAttribute>();
+        if (isConnected(srcAttr, dstUsdAttr)) {
+            return false;
+        }
+    }
+
+    PXR_NS::UsdShadeNodeGraph ngPrim(prim);
+
+    if (!ngPrim) {
+        return true;
+    }
+
+    // Do not remove if there is a connection with a child prim.
```

Comment:
There is a very useful workflow where you drag the first port of the type you want on the boundary, and connect it to the input node as many times as desired, then delete the connections.
In this scenario, the user expects the ports on the input node to remain so he can do further work on them. He can always delete the ones he does not want anymore.
That would mean returning false here even if there are no connections left. Same thing on the reverse function.

Reply:
Thank you. I didn't know about this workflow. 
I've updated the solution also with the test.
 
I've also reintroduced the condition for the `UsdShadeMaterial` which we should keep ( it is also required by some tests).